language/move-prover/tests/sources/functional/generic_invariants.exp

```diff
@@ -147,12 +146,32 @@ error: global memory invariant does not hold
    =     at tests/sources/functional/generic_invariants.move:39
    =     at tests/sources/functional/generic_invariants.move:44
 
+error: global memory invariant does not hold
```

Comment:
It is interesting that the new translation scheme produces *more* errors; the same invariant is failing multiple times (I guess for the generic and instantiated versions). But I put both generated bytecode and boogie code SxS and there are really no semantic differences (the old code created `assume/assert true` on Boogie level -- and the order in which instantiated functions are generated is different. Butterfly? We ask Boogie to generate as much errors as it can.

Reply:
I suspect that [this](https://github.com/diem/diem/pull/8946/files#diff-286c8224b92dcae98ad2de9ea2621374d3e798ba9e56bf1d2e22906d68dde48aR520) is the reason that causes the problem. The related invariant set calculated in the `analysis` phase of the `global_invariant_instrumentation.rs` pass does not play well with the `InvariantAnalysisData` calculated in the `verification_analysis.rs` pass. (In fact, the related invariant set derived from the analysis phase is ignored in the instrumentation phase, which is not good).

Reconciling them is not in the scope of this PR, so I am in favor of landing it first and work on reconciling them later. But we should not let the gap be open for too long.

Reply:
The link in the diff unfortunately does not work (they never do), so not sure what you spotted. There is no change in this PR regards the `related_invariants`, which is and was only used to filter entrypoint_invariants once. Also recall that I diffed the output. Here is the [diff of the bytecode of the end of the pipeline](https://editor.mergely.com/axRd5vZG/)  for generic_invariants.move (lhs before, rhs after). I cleaned up the ordering problem manually, and AFAICT the only difference is now the `assume/assert true` which is not longer in the output.

Reply:
Thanks for the diff Wolfgang!

What I observe is that some invariants that I don't expect to be instrumented ... gets instrumented, and this happens both before and after this PR so this is unlikely the reason why we now have more errors from Boogie,

An example is line 248 on the RHS in the diff:
I don't expect the following invariant
```
    invariant
        exists<S::Storage<u64, bool>>(@0x22)
            ==> global<S::Storage<u64, bool>>(@0x22).x == 1;
```
To be applied to the fully generic version of `public_x_y`. It does not hurt to instrument it there, but based on my reading of the code, that invariant should only be instrumented in the instantiated version of `public_x_y [X := u64, Y := bool]`.